### PR TITLE
Human readable sizes in logs

### PR DESCRIPTION
**Pull request by Spaceman1984**
## Description

This PR adds outputting human readable byte sizes in the management server logs, agent logs, and usage records. A non-dynamic global variable is added (display.human.readable.sizes) to control switching this feature on and off. This setting is sent to the agent on connection and is only read from the database when the management server is started up. The setting is kept in memory by the use of a static field on the NumbersUtil class and is available throughout the codebase.

Instead of seeing things like: 
2020-07-23 15:31:58,593 DEBUG [c.c.a.t.Request] (AgentManager-Handler-12:null) (logid:) Seq 8-1863645820801253428: Processing:  { Ans: , MgmtId: 52238089807, via: 8, Ver: v1, Flags: 10, [{"com.cloud.agent.api.NetworkUsageAnswer":{"routerName":"r-224-VM",**"bytesSent":"106496"**,**"bytesReceived":"0"**,"result":"true","details":"","wait":"0",}}] }

The KB MB and GB values will be printed out:

2020-07-23 15:31:58,593 DEBUG [c.c.a.t.Request] (AgentManager-Handler-12:null) (logid:) Seq 8-1863645820801253428: Processing:  { Ans: , MgmtId: 52238089807, via: 8, Ver: v1, Flags: 10, [{"com.cloud.agent.api.NetworkUsageAnswer":{"routerName":"r-224-VM",**"bytesSent":"(104.00 KB) 106496"**,**"bytesReceived":"(0 bytes) 0"**,"result":"true","details":"","wait":"0",}}] }

FS: https://cwiki.apache.org/confluence/display/CLOUDSTACK/Human+Readable+Byte+sizes

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

This has been tested by observing the logs on the management server and on the agent while doing things like the following (Also inspecting the events table and cloud_usage table after the usage aggregation job has run.):

* Upload an iso.
* Deploy a vm.
* Start a vm.
* Stop a vm.
* Migrate a vm.
* Add a storage volume to a zone.
* Resize the volume.

Tested on a VMWare, Xenserver and a KVM cluster.

### CODE REVIEW

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
@Spaceman1984 is this PR ready for review?

**Comment by Spaceman1984** [comment]
> @Spaceman1984 is this PR ready for review?

@GabrielBrascher not yet, just a few more tweaks.

**Comment by Spaceman1984** [comment]
@blueorangutan test

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: core/src/main/java/com/cloud/agent/transport/Request.java [hunk 1]
```diff
@@ -395,7 +396,7 @@ public String toString() {
         return log("", true, Level.DEBUG);
     }
 
-    protected String log(String msg, boolean logContent, Level level) {
+    protected String  log(String msg, boolean logContent, Level level) {
```
Comment (DaanHoogland) [inline core/src/main/java/com/cloud/agent/transport/Request.java:399]: extra space

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@Spaceman1984 failures related to vm lifecycle seen here as well, I suspect it may be related to the fix you added in #4226

File: utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    @Test
+    public void parseComplexJsonTest() {
+        assertEquals("[{\"object\":[]}]", getHumanReadableBytesJson("[{\"object\":[]}]"));
+        assertEquals("[{\"object\":[{},{}]}]", getHumanReadableBytesJson("[{\"object\":[{},{}]}]"));
+        assertEquals("[{\"object\":[]},{\"object\":[]}]", getHumanReadableBytesJson("[{\"object\":[]},{\"object\":[]}]"));
+        assertEquals("[{\"object\":[{\"object\":[]}]},{\"object\":[]}]", getHumanReadableBytesJson("[{\"object\":[{\"object\":[]}]},{\"object\":[]}]"));
+    }
+    @Test
+    public void parseMatchJsonTest() {
```
Comment (rhtyd) [inline utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java:49]: @Spaceman1984 can you add tests for non zero values, and see if the expected results match? (for kbs, mbs, gbs)

File: utils/src/main/java/com/cloud/utils/NumbersUtil.java [hunk 3]
```diff
@@ -93,6 +95,13 @@ public static String toReadableSize(long bytes) {
         return builder.toString();
     }
 
+    public static String toHumanReadableSize(long size) {
```
Comment (rhtyd) [inline utils/src/main/java/com/cloud/utils/NumbersUtil.java:98]: @Spaceman1984 can you add unit test for this method, giving some values such as zero and non-zero (to test KB, MB, GB, TB). Have you verified that all consumes would pass `long` to it and not `Long` which may potentially cause a NPE?

**Comment by rhtyd** [comment]
@Spaceman1984 overall LGTM, only throughout we'll know the extent of what was missed. I've left some comments and advised few test cases that we can add, please address them.
Also check the human readable json converter that in the original example at the top should say `104.0KB` for `"bytesSent":"106496",` value... please check that as well.

**Comment by Spaceman1984** [comment]
@rhtyd, I checked the commas vs dots in the conversion, looks like it changes based on the locale and JDK version.

File: utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java (hunk 2 above)
Comment (Spaceman1984) [reply utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java:50]: @rhtyd these tests already exist in NumbersUtilTest.java

File: utils/src/main/java/com/cloud/utils/NumbersUtil.java (hunk 3 above)
Comment (Spaceman1984) [reply utils/src/main/java/com/cloud/utils/NumbersUtil.java:103]: I have added a check and a test for null which prevents an NPE.

File: utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java [hunk 4]
```diff
@@ -50,4 +51,14 @@ public void parseMatchJsonTest() {
         assertEquals("[{\"size\":\"(0 bytes) 0\"}]", getHumanReadableBytesJson("[{\"size\": \"0\"}]"));
         assertEquals("[{\"size\":\"(0 bytes) 0\",\"bytesSent\":\"(0 bytes) 0\"}]", getHumanReadableBytesJson("[{\"size\": \"0\", \"bytesSent\": \"0\"}]"));
     }
+
+    @Test
+    public void localeTest() {
+        Locale.setDefault(Locale.UK); // UK test
```
Comment (rhtyd) [inline utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java:57]: @Spaceman1984 what's the default locale for CloudStack, is there a global setting or does it pick it up from OS?

Comment (Spaceman1984) [reply utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java:57]: @rhtyd From Java Docs:
"The Java Virtual Machine sets the default locale during startup based on the host environment. It is used by many locale-sensitive methods if no locale is explicitly specified. It can be changed using the setDefault method."

**Comment by rhtyd** [comment]
@blueorangutan package

File: test/integration/smoke/test_certauthority_root.py [hunk 5]
```diff
@@ -16,11 +16,10 @@
...
-from marvin.cloudstackTestCase import *
-from marvin.cloudstackAPI import *
-from marvin.lib.utils import *
+from marvin.cloudstackTestCase import cloudstackTestCase
+from marvin.lib.utils import cleanup_resources
 from marvin.lib.base import *
-from marvin.lib.common import *
+from marvin.lib.common import list_hosts
 
 from cryptography import x509
```
Comment (borisstoyanov) [inline test/integration/smoke/test_certauthority_root.py:24]: @Spaceman1984 why do we need changes in this file?

Comment (Spaceman1984) [reply test/integration/smoke/test_certauthority_root.py:24]: @borisstoyanov tests were failing when run outside of Travis.

File: utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java (hunk 4 above)
Comment (DaanHoogland) [reply utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java:57]: I think "C" is the default locale. maybe this or some other more worldly locale is a good idea?

**Comment by rhtyd** [comment]
@Spaceman1984 can you fix the conflicts and address any outstanding comments. Thanks.

File: utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java (hunk 4 above)
Comment (Spaceman1984) [reply utils/src/test/java/com/cloud/utils/HumanReadableJsonTest.java:59]: @DaanHoogland I searched the code base, Cloudstack doesn't set a default locale, it will be picked up from the environment, which I believe is the correct behaviour.

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

**Comment by alexandremattioli** [comment]
LGTM

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Tests LGTM, I'll merge this after few other PRs as this could potentially cause merge conflict with other PRs cc @DaanHoogland
